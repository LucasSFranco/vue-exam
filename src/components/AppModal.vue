<template>

  <slot :open="open" />

  <teleport to="body">
    <div
      v-if="isVisible"
      class="modal"
      @mousedown.self="close"
    >
      <div class="modal__content">
        <button
          class="modal__closer"
          @click="close"
        >
          <Icon :icon="['fas', 'times']" />
        </button>

        <slot name="content" />

      </div>
    </div>
  </teleport>
</template>

<script>

export default {
  name: 'AppModal',
  data() {
    return {
      isVisible: false,
    }
  },
  methods: {
    open() {
      this.isVisible = true
    },
    close() {
      this.isVisible = false
    },
  },
}

</script>

<style lang="sass" scoped>

  .modal
    display: flex
    align-items: center
    justify-content: center

    position: absolute
    top: 0
    left: 0
    right: 0
    bottom: 0

    background: rgba(0, 0, 0, .25)
    padding: 1rem

    z-index: 999

    .modal__content
      position: relative

      .modal__closer
        position: absolute
        top: .75rem
        right: .75rem

        color: var(--text-muted)
        cursor: pointer
        font-size: 1.25rem

        transition: 50ms color

        z-index: 99

        &:hover
          color: var(--text-normal)

      h3
        color: var(--header-primary)
        font-size: 1.5rem

      a
        color: var(--text-link)
        font-family: 'Roboto', sans-serif
        font-size: .75rem

        &:hover
          text-decoration: underline

      p
        color: var(--text-normal)
        font-size: 1rem
        margin-top: 1.5rem

</style>
