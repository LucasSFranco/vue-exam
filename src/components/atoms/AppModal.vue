<template>

  <slot
    name="trigger"
    :open="open"
  />

  <teleport to="body">
    <div
      v-if="isVisible"
      class="modal"
      @mousedown.self="close"
    >
      <div class="modal__content">
        <button
          class="modal__closer"
          @click="close"
        >
          <Icon icon="times" />
        </button>

        <slot name="content" />

      </div>
    </div>
  </teleport>
</template>

<script>

export default {
  name: 'AppModal',
  data() {
    return {
      isVisible: false,
    }
  },
  methods: {
    open() {
      this.isVisible = true
    },
    close() {
      this.isVisible = false
    },
  },
}

</script>

<style lang="scss" scoped>

  .modal {
    display: flex;
    align-items: center;
    justify-content: center;

    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;

    background: rgba(0, 0, 0, .25);
    z-index: 999;

    .modal__content {
      position: relative;

      .modal__closer {
        position: absolute;
        top: .75rem;
        right: .75rem;

        color: var(--text-secondary);
        cursor: pointer;
        font-size: 1.25rem;
        z-index: 99;

        transition: 50ms color;

        &:hover {
          color: var(--text-primary);
        }
      }
    }
  }

</style>
