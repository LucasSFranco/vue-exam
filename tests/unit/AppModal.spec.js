import { shallowMount } from '@vue/test-utils'

import AppModal from '@/components/atoms/AppModal.vue'

describe('Modal', () => {
  it('properly renders snapshot', () => {
    const wrapper = shallowMount(AppModal, {
      slots: {
        trigger: 'Modal Trigger',
        content: 'Modal Content'
      },
      global: {
        stubs: {
          Icon: true
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
