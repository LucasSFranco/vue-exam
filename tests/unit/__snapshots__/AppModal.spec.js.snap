// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal properly renders snapshot 1`] = `"Modal Trigger<stub></stub>"`;
